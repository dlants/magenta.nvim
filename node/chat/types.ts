export type Role = "user" | "assistant";

export type ThreadId = number & { __threadId: true };

export type ThreadType =
  | "subagent_learn"
  | "subagent_plan"
  | "subagent_default"
  | "subagent_fast"
  | "root";
