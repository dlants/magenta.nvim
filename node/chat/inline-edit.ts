import type { Nvim } from "nvim-node";
import { NvimBuffer, type BufNr, type Line } from "../nvim/buffer";
import type { WindowId } from "../nvim/window";
import { assertUnreachable } from "../utils/assertUnreachable";
import type { Dispatch, Update } from "../tea/tea";
import { getCurrentWindow } from "../nvim/nvim";
import type { InlineEditToolRequest } from "../tools/inlineEdit";
import type { Result } from "../utils/result";
import type { StopReason, Usage } from "../providers/provider";

export type Model =
  | {
      state: "inactive";
    }
  | {
      state: "error";
      error: string;
    }
  | {
      state: "awaiting-prompt";
      targetWindowId: WindowId;
      targetBufnr: BufNr;
      inputWindowId: WindowId;
      inputBufnr: BufNr;
    }
  | {
      state: "response-pending";
      targetWindowId: WindowId;
      targetBufnr: BufNr;
      inputWindowId: WindowId;
      inputBufnr: BufNr;
    };

export type Msg =
  | {
      type: "start-inline-edit";
    }
  | {
      type: "error";
      error: string;
    }
  | {
      type: "inline-edit-initialized";
      targetWindowId: WindowId;
      targetBufnr: BufNr;
      inputWindowId: WindowId;
      inputBufnr: BufNr;
    }
  | {
      type: "submit-inline-edit";
    }
  | {
      type: "inline-edit-tool-request";
      inlineEdit: Result<InlineEditToolRequest, { rawRequest: unknown }>;
      stopReason: StopReason;
      usage: Usage;
    }
  | {
      type: "dismiss-inline-edit";
    };

export function init({ nvim }: { nvim: Nvim }) {
  function initModel(): Model {
    return {
      state: "inactive",
    };
  }

  const update: Update<Msg, Model> = (msg, model) => {
    switch (msg.type) {
      case "start-inline-edit":
        return [model, startInlineEdit];

      case "inline-edit-initialized":
        return [
          {
            state: "awaiting-prompt",
            targetWindowId: msg.targetWindowId,
            targetBufnr: msg.targetBufnr,
            inputWindowId: msg.inputWindowId,
            inputBufnr: msg.inputBufnr,
          },
        ];

      case "submit-inline-edit":
        if (model.state === "awaiting-prompt") {
          return [
            {
              state: "response-pending",
              targetWindowId: model.targetWindowId,
              targetBufnr: model.targetBufnr,
              inputWindowId: model.inputWindowId,
              inputBufnr: model.inputBufnr,
            },
            // thunk will be handled by parent/chat
          ];
        } else {
          return [{ state: "inactive" }];
        }

      case "dismiss-inline-edit":
        return [
          {
            state: "inactive",
          },
        ];

      case "error":
        return [{ state: "error", error: msg.error }];

      case "inline-edit-tool-request": {
        if (model.state != "response-pending") {
          return [
            { state: "error", error: `Unexpected state: ${model.state}` },
          ];
        }
        if (msg.inlineEdit.status === "error") {
          return [
            {
              state: "error",
              error: `inlineEdit error: ${msg.inlineEdit.error}\nrawRequest:${JSON.stringify(msg.inlineEdit.rawRequest, null, 2)}`,
            },
          ];
        }

        const input = msg.inlineEdit.value.input;

        return [
          { state: "inactive" },
          async (dispatch) => {
            const { targetWindowId } = model;

            const targetBufnr = (await nvim.call("nvim_win_get_buf", [
              targetWindowId,
            ])) as BufNr;
            const buffer = new NvimBuffer(targetBufnr, nvim);
            const lines = await buffer.getLines({ start: 0, end: -1 });
            const content = lines.join("\n");

            const replaceStart = content.indexOf(input.find);
            if (replaceStart === -1) {
              dispatch({
                type: "error",
                error: `Unable to find text in buffer:
\`\`\`
${input.find}
\`\`\``,
              });
              return;
            }
            const replaceEnd = replaceStart + input.find.length;

            const nextContent =
              content.slice(0, replaceStart) +
              "\n>>>>>>> Suggested change\n" +
              input.replace +
              "\n=======\n" +
              input.find +
              "\n<<<<<<< Current\n" +
              content.slice(replaceEnd);

            await buffer.setLines({
              start: 0,
              end: -1,
              lines: nextContent.split("\n") as Line[],
            });
          },
        ];
      }

      default:
        assertUnreachable(msg);
    }
  };

  async function startInlineEdit(dispatch: Dispatch<Msg>) {
    const targetWindow = await getCurrentWindow(nvim);
    const isMagentaWindow = await targetWindow.getVar("magenta");

    if (!isMagentaWindow) {
      const targetBufnr = (await nvim.call("nvim_win_get_buf", [
        targetWindow.id,
      ])) as BufNr;

      const scratchBuffer = await NvimBuffer.create(false, true, nvim);
      await scratchBuffer.setOption("bufhidden", "wipe");

      const inlineInputWindowId = (await nvim.call("nvim_open_win", [
        scratchBuffer.id,
        true, // enter the input window
        {
          win: targetWindow.id, // split inside current window
          split: "above",
          height: 10,
          style: "minimal",
        },
      ])) as WindowId;

      // Set up <CR> mapping in normal mode
      await scratchBuffer.setKeymap({
        mode: "n",
        lhs: "<CR>",
        rhs: ":Magenta submit-inline-edit<CR>",
        opts: { silent: true, noremap: true },
      });

      dispatch({
        type: "inline-edit-initialized",
        targetWindowId: targetWindow.id,
        targetBufnr,
        inputWindowId: inlineInputWindowId,
        inputBufnr: scratchBuffer.id,
      });
    }
  }

  return { initModel, update };
}
