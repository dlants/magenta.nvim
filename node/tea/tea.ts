import {
  d,
  type MountedVDOM,
  type MountedView,
  type MountPoint,
  mountView,
  type VDOMNode,
} from "./view.ts";
import { BINDING_KEYS, type BindingKey, getBindings } from "./bindings.ts";
import { getCurrentWindow, notifyErr } from "../nvim/nvim.ts";
import type { Row0Indexed } from "../nvim/window.ts";
import type { Nvim } from "nvim-node";
import { Defer } from "../utils/async.ts";

export type Dispatch<Msg> = (msg: Msg) => void;

export type Update<Msg, Model, Context = undefined> = Context extends undefined
  ? (msg: Msg, model: Model) => [Model] | [Model, Thunk<Msg> | undefined]
  : (
      msg: Msg,
      model: Model,
      context: Context,
    ) => [Model] | [Model, Thunk<Msg> | undefined];

export type View<Model> = ({ model }: { model: Model }) => VDOMNode;

type AppState<Model> =
  | {
      status: "running";
      model: Model;
    }
  | {
      status: "error";
      error: string;
    };

export type MountedApp = {
  onKey(key: BindingKey): void;
  render(): void;
  unmount(): void;
  getMountedNode(): MountedVDOM;
  waitForRender(): Promise<void>;
};

export type App<Model> = {
  mount(mount: MountPoint): Promise<MountedApp>;
  getState(): AppState<Model>;
  destroy(): void;
};

export function createApp<Model>({
  nvim,
  initialModel,
  View,
}: {
  nvim: Nvim;
  initialModel: Model;
  View: View<Model>;
}): App<Model> {
  let currentState: AppState<Model> = {
    status: "running",
    model: initialModel,
  };
  let root: MountedView<{ currentState: AppState<Model> }> | undefined;

  let renderDefer: Defer<void> | undefined;
  let renderPromise: Promise<void> | undefined;
  let reRender = false;

  function render() {
    if (renderPromise) {
      reRender = true;
    } else {
      if (!renderDefer) {
        renderDefer = new Defer();
      }

      if (root) {
        renderPromise = root
          .render({ currentState })
          .catch((err) => {
            nvim.logger?.error(err as Error);
            // eslint-disable-next-line @typescript-eslint/no-floating-promises
            notifyErr(nvim, "render", err);
            if (renderDefer) {
              renderDefer.reject(err as Error);
              renderDefer = undefined;
            }
          })
          .finally(() => {
            renderPromise = undefined;
            if (reRender) {
              reRender = false;
              nvim.logger?.debug(`followup render triggered`);
              render();
            } else {
              if (renderDefer) {
                renderDefer.resolve();
                renderDefer = undefined;
              }
            }
          });
      }
    }
  }

  function App({ currentState }: { currentState: AppState<Model> }) {
    return d`${
      currentState.status == "running"
        ? View({ model: currentState.model })
        : d`Error: ${currentState.error}`
    }`;
  }

  return {
    async mount(mount: MountPoint) {
      root = await mountView({
        view: App,
        mount,
        props: { currentState },
      });

      for (const vimKey of BINDING_KEYS) {
        try {
          await nvim.call("nvim_exec_lua", [
            `require('magenta').listenToBufKey(${mount.buffer.id}, "${vimKey}")`,
            [],
          ]);
        } catch (e) {
          throw new Error(`failed to nvim_exec_lua: ${JSON.stringify(e)}`);
        }
      }

      return {
        getMountedNode() {
          return root!._getMountedNode();
        },

        unmount() {
          if (root) {
            root.unmount();
            root = undefined;
          }
        },

        render() {
          render();
        },

        async waitForRender() {
          if (renderDefer) {
            await renderDefer.promise;
          }
        },

        async onKey(key: BindingKey) {
          const window = await getCurrentWindow(mount.nvim);
          const buffer = await window.buffer();
          if (buffer.id != mount.buffer.id) {
            nvim.logger?.warn(
              `Got onKey event ${key}, but current window is not showing mounted buffer`,
            );
            return;
          }
          const { row, col } = await window.getCursor();
          if (root) {
            // win_get_cursor is 1-indexed, while our positions are 0-indexed
            const bindings = getBindings(root._getMountedNode(), {
              row: (row - 1) as Row0Indexed,
              col,
            });
            if (bindings && bindings[key]) {
              bindings[key]();
            }
          } else {
            nvim.logger?.debug(
              `Got onKey event ${key}, but root is no longer mounted.`,
            );
          }
        },
      };
    },
    getState() {
      return currentState;
    },
    destroy() {
      if (root) {
        root.unmount();
        root = undefined;
      }
      currentState = {
        status: "error",
        error: "destroyed",
      };
    },
  };
}

export type Thunk<Msg> = (dispatch: Dispatch<Msg>) => Promise<void>;

export function wrapThunk<MsgType extends string, InnerMsg>(
  msgType: MsgType,
  thunk: Thunk<InnerMsg> | undefined,
): Thunk<{ type: MsgType; msg: InnerMsg }> | undefined {
  if (!thunk) {
    return undefined;
  }
  return (dispatch: Dispatch<{ type: MsgType; msg: InnerMsg }>) =>
    thunk((msg: InnerMsg) => dispatch({ type: msgType, msg }));
}

export function chainThunks<Msg>(
  ...thunks: (Thunk<Msg> | undefined)[]
): Thunk<Msg> {
  return async (dispatch) => {
    for (const thunk of thunks) {
      if (thunk) {
        await thunk(dispatch);
      }
    }
  };
}

export function parallelThunks<Msg>(
  ...thunks: (Thunk<Msg> | undefined)[]
): Thunk<Msg> {
  return async (dispatch) => {
    await Promise.all(thunks.map((t) => t && t(dispatch)));
  };
}
