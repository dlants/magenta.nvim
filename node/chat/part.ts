import {
  ToolManager,
  type Msg as ToolManagerMsg,
  type ToolRequestId,
} from "../tools/toolManager.ts";
import { assertUnreachable } from "../utils/assertUnreachable.ts";
import { d, type View } from "../tea/view.ts";
import type {
  ProviderMessageContent,
  ProviderToolResultContent,
  StopReason,
  Usage,
} from "../providers/provider.ts";
import { wrapThunk, type Dispatch, type Thunk } from "../tea/tea.ts";

type State =
  | {
      type: "text";
      text: string;
    }
  | {
      type: "tool-request";
      requestId: ToolRequestId;
    }
  | {
      type: "malformed-tool-request";
      error: string;
      rawRequest: unknown;
    }
  | {
      type: "stop-msg";
      stopReason: StopReason;
      usage: Usage;
    };

export type Msg = {
  type: "tool-manager-msg";
  msg: ToolManagerMsg;
};

export class Part {
  toolManager: ToolManager;
  state: State;

  constructor({
    state,
    toolManager,
  }: {
    state: State;
    toolManager: ToolManager;
  }) {
    this.state = state;
    this.toolManager = toolManager;
  }

  update(msg: Msg): Thunk<Msg> | undefined {
    switch (msg.type) {
      case "tool-manager-msg": {
        const thunk = this.toolManager.update(msg.msg);
        return wrapThunk("tool-manager-msg", thunk);
      }
      default:
        return assertUnreachable(msg.type);
    }
  }

  toMessageContent(): {
    content?: ProviderMessageContent;
    result?: ProviderToolResultContent;
  } {
    switch (this.state.type) {
      case "text":
        return { content: this.state };

      case "tool-request": {
        const toolWrapper =
          this.toolManager.state.toolWrappers[this.state.requestId];
        return {
          content: {
            type: "tool_use",
            request: toolWrapper.tool.request,
          },
          result: toolWrapper.tool.getToolResult(),
        };
      }

      case "malformed-tool-request": {
        return {
          content: {
            type: "text",
            text: `Malformed tool request: ${this.state.error}`,
          },
        };
      }

      case "stop-msg": {
        return {};
      }

      default:
        return assertUnreachable(this.state);
    }
  }
}

export const view: View<{
  part: Part;
  dispatch: Dispatch<Msg>;
}> = ({ part, dispatch }) => {
  switch (part.state.type) {
    case "text":
      return d`${part.state.text}`;

    case "malformed-tool-request":
      return d`Malformed tool request: ${part.state.error}
${JSON.stringify(part.state.rawRequest, null, 2) || "undefined"}`;

    case "tool-request": {
      const toolWrapper =
        part.toolManager.state.toolWrappers[part.state.requestId];
      if (!toolWrapper) {
        throw new Error(
          `Unable to find model with requestId ${part.state.requestId}`,
        );
      }
      return part.toolManager.renderTool(toolWrapper, (msg) =>
        dispatch({
          type: "tool-manager-msg",
          msg,
        }),
      );
    }

    case "stop-msg": {
      return d`Stopped (${part.state.stopReason}) [input: ${part.state.usage.inputTokens.toString()}, output: ${part.state.usage.outputTokens.toString()}${
        part.state.usage.cacheHits !== undefined
          ? d`, cache hits: ${part.state.usage.cacheHits.toString()}`
          : ""
      }${
        part.state.usage.cacheMisses !== undefined
          ? d`, cache misses: ${part.state.usage.cacheMisses.toString()}`
          : ""
      }]`;
    }

    default:
      assertUnreachable(part.state);
  }
};
