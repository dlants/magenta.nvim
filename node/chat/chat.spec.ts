import {
  extractMountTree,
  withNvimClient,
  withDriver,
} from "../test/preamble.ts";
import * as Chat from "./chat.ts";
import { type ToolRequestId } from "../tools/toolManager.ts";
import { createApp } from "../tea/tea.ts";
import { describe, expect, it } from "vitest";
import { pos } from "../tea/view.ts";
import { NvimBuffer, type Line } from "../nvim/buffer.ts";
import type { MessageId } from "./message.ts";
import * as ListDirectory from "../tools/listDirectory.ts";
import * as ListBuffers from "../tools/listBuffers.ts";

describe("tea/chat.spec.ts", () => {
  it("chat render and a few updates", async () => {
    await withNvimClient(async (nvim) => {
      const buffer = await NvimBuffer.create(false, true, nvim);
      await buffer.setOption("modifiable", false);
      const chatModel = Chat.init({
        nvim,
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-explicit-any
        lsp: undefined as any,
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-explicit-any
        options: undefined as any,
      });
      const model = chatModel.initModel({
        name: "claude-3-7",
        provider: "anthropic",
        model: "claude-3-7-latest",
      });

      const app = createApp({
        nvim,
        initialModel: model,
        update: (model, msg) =>
          // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-explicit-any
          chatModel.update(model, msg, { nvim, options: undefined as any }),
        View: chatModel.view,
        suppressThunks: true,
      });

      const mountedApp = await app.mount({
        nvim,
        buffer,
        startPos: pos(0, 0),
        endPos: pos(-1, -1),
      });

      await mountedApp.waitForRender();

      expect(
        await buffer.getLines({ start: 0, end: -1 }),
        "initial render of chat works",
      ).toEqual(Chat.LOGO.split("\n") as Line[]);

      app.dispatch({
        type: "add-message",
        role: "user",
        content: "Can you look at my list of buffers?",
      });
      await mountedApp.waitForRender();

      app.dispatch({
        type: "stream-response",
        text: "Sure, let me use the list_buffers tool.",
      });
      await mountedApp.waitForRender();

      app.dispatch({
        type: "init-tool-use",
        request: {
          status: "ok",
          value: {
            id: "request-id" as ToolRequestId,
            input: {},
            name: "list_buffers",
          },
        },
      });
      await mountedApp.waitForRender();

      expect(
        await buffer.getLines({ start: 0, end: -1 }),
        "in-progress render is as expected",
      ).toEqual([
        "# user:",
        "Can you look at my list of buffers?",
        "",
        "# assistant:",
        "Sure, let me use the list_buffers tool.",
        "⚙️ Grabbing buffers...",
        "",
        "Stopped (end_turn)",
      ] as Line[]);

      expect(
        await extractMountTree(mountedApp.getMountedNode()),
      ).toMatchSnapshot();

      app.dispatch({
        type: "tool-manager-msg",
        msg: {
          type: "tool-msg",
          id: "request-id" as ToolRequestId,
          msg: {
            type: "list_buffers",
            msg: {
              type: "finish",
              result: {
                status: "ok",
                value: "some buffer content",
              },
            },
          },
        },
      });
      await mountedApp.waitForRender();

      expect(
        await buffer.getLines({ start: 0, end: -1 }),
        "finished render is as expected",
      ).toEqual([
        "# user:",
        "Can you look at my list of buffers?",
        "",
        "# assistant:",
        "Sure, let me use the list_buffers tool.",
        "✅ Finished getting buffers.",
        "",
        "Stopped (end_turn)",
      ] as Line[]);
    });
  });

  it("chat clear", async () => {
    await withNvimClient(async (nvim) => {
      const buffer = await NvimBuffer.create(false, true, nvim);
      await buffer.setOption("modifiable", false);
      const chatModel = Chat.init({
        nvim,
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-explicit-any
        lsp: undefined as any,
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-explicit-any
        options: undefined as any,
      });
      const model = chatModel.initModel({
        name: "claude-3-7",
        provider: "anthropic",
        model: "claude-3-7-latest",
      });

      const app = createApp({
        nvim,
        initialModel: model,
        update: (model, msg) =>
          // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-explicit-any
          chatModel.update(model, msg, { nvim, options: undefined as any }),
        View: chatModel.view,
        suppressThunks: true,
      });

      const mountedApp = await app.mount({
        nvim,
        buffer,
        startPos: pos(0, 0),
        endPos: pos(-1, -1),
      });

      await mountedApp.waitForRender();

      expect(
        await buffer.getLines({ start: 0, end: -1 }),
        "initial render of chat works",
      ).toEqual(Chat.LOGO.split("\n") as Line[]);

      app.dispatch({
        type: "add-message",
        role: "user",
        content: "Can you look at my list of buffers?",
      });
      await mountedApp.waitForRender();

      app.dispatch({
        type: "stream-response",
        text: "Sure, let me use the list_buffers tool.",
      });
      await mountedApp.waitForRender();

      expect(
        await buffer.getLines({ start: 0, end: -1 }),
        "in-progress render is as expected",
      ).toEqual([
        "# user:",
        "Can you look at my list of buffers?",
        "",
        "# assistant:",
        "Sure, let me use the list_buffers tool.",
        "",
        "Stopped (end_turn)",
      ] as Line[]);

      app.dispatch({
        type: "clear",
        profile: {
          name: "claude-3-7",
          provider: "anthropic",
          model: "claude-3-7-latest",
        },
      });
      await mountedApp.waitForRender();

      expect(
        await buffer.getLines({ start: 0, end: -1 }),
        "finished render is as expected",
      ).toEqual(Chat.LOGO.split("\n") as Line[]);
    });
  });

  it("getMessages correctly interleaves tool requests and responses", async () => {
    await withNvimClient(async (nvim) => {
      const chatModel = Chat.init({
        nvim,
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-explicit-any
        lsp: undefined as any,
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-explicit-any
        options: undefined as any,
      });
      const model = chatModel.initModel({
        name: "claude-3-7",
        provider: "anthropic",
        model: "claude-3-7-latest",
      });

      model.messages.push({
        id: 1 as MessageId,
        role: "user",
        parts: [{ type: "text", text: "Can you help me with my code?" }],
        edits: {},
      });

      const TOOL1_ID = "tool-1" as ToolRequestId;
      const TOOL2_ID = "tool-2" as ToolRequestId;

      model.messages.push({
        id: 2 as MessageId,
        role: "assistant",
        parts: [
          {
            type: "text",
            text: "I'll help you. Let me check your files first.",
          },
          {
            type: "tool-request",
            requestId: TOOL1_ID,
          },
          { type: "text", text: "Now let me check your buffers too." },
          {
            type: "tool-request",
            requestId: TOOL2_ID,
          },
          { type: "text", text: "Based on these results, I can help you." },
        ],
        edits: {},
      });

      {
        const toolModel = ListDirectory.initModel(
          {
            id: TOOL1_ID,
            name: "list_directory",
            input: {},
          },
          { nvim },
        )[0];
        toolModel.state = {
          state: "done",
          result: {
            type: "tool_result",
            id: TOOL1_ID,
            result: {
              status: "ok",
              value: "list_directory result",
            },
          },
        };

        model.toolManager.toolWrappers[TOOL1_ID] = {
          model: toolModel,
          showRequest: true,
          showResult: true,
        };
      }

      {
        const toolModel = ListBuffers.initModel(
          {
            id: TOOL2_ID,
            name: "list_buffers",
            input: {},
          },
          { nvim },
        )[0];

        toolModel.state = {
          state: "done",
          result: {
            type: "tool_result",
            id: TOOL2_ID,
            result: {
              status: "ok",
              value: "list_buffers result",
            },
          },
        };

        model.toolManager.toolWrappers[TOOL2_ID] = {
          model: toolModel,
          showRequest: true,
          showResult: true,
        };
      }

      const messages = await chatModel.getMessages(model);

      expect(messages).toEqual([
        {
          role: "user",
          content: [{ type: "text", text: "Can you help me with my code?" }],
        },
        {
          role: "assistant",
          content: [
            {
              type: "text",
              text: "I'll help you. Let me check your files first.",
            },
            {
              type: "tool_use",
              request: {
                id: "tool-1",
                input: {},
                name: "list_directory",
              },
            },
          ],
        },
        {
          role: "user",
          content: [
            {
              type: "tool_result",
              id: "tool-1",
              result: {
                status: "ok",
                value: "list_directory result",
              },
            },
          ],
        },
        {
          role: "assistant",
          content: [
            { type: "text", text: "Now let me check your buffers too." },
            {
              type: "tool_use",
              request: {
                id: "tool-2",
                input: {},
                name: "list_buffers",
              },
            },
          ],
        },
        {
          role: "user",
          content: [
            {
              type: "tool_result",
              id: "tool-2",
              result: {
                status: "ok",
                value: "list_buffers result",
              },
            },
          ],
        },
        {
          role: "assistant",
          content: [
            { type: "text", text: "Based on these results, I can help you." },
          ],
        },
      ]);
    });
  });

  it("handles errors during streaming response", async () => {
    await withDriver(async (driver) => {
      await driver.showSidebar();

      const userPrompt = "Generate some text for me";
      await driver.inputMagentaText(userPrompt);
      await driver.send();

      const pendingRequest = await driver.mockAnthropic.awaitPendingRequest();

      pendingRequest.onText("I'm generating text for you...");

      await driver.assertDisplayBufferContains(
        "I'm generating text for you...",
      );

      pendingRequest.defer.reject(
        new Error("Connection error during response"),
      );

      await driver.assertDisplayBufferContains(
        "Error Connection error during response",
      );

      // Check that the input buffer is pre-populated with the last user message
      await driver.assertInputBufferContains(userPrompt);
    });
  });
});
