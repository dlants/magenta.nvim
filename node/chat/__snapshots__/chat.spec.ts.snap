// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tea/chat.spec.ts > chat render and a few updates 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "children": [
                    {
                      "content": "# ",
                      "endPos": {
                        "col": 2,
                        "row": 0,
                      },
                      "startPos": {
                        "col": 0,
                        "row": 0,
                      },
                      "type": "string",
                    },
                    {
                      "content": "user",
                      "endPos": {
                        "col": 6,
                        "row": 0,
                      },
                      "startPos": {
                        "col": 2,
                        "row": 0,
                      },
                      "type": "string",
                    },
                    {
                      "content": ":
",
                      "endPos": {
                        "col": 0,
                        "row": 1,
                      },
                      "startPos": {
                        "col": 6,
                        "row": 0,
                      },
                      "type": "string",
                    },
                    {
                      "children": [
                        {
                          "children": [
                            {
                              "children": [
                                {
                                  "content": "Can you look at my list of buffers?",
                                  "endPos": {
                                    "col": 35,
                                    "row": 1,
                                  },
                                  "startPos": {
                                    "col": 0,
                                    "row": 1,
                                  },
                                  "type": "string",
                                },
                              ],
                              "endPos": {
                                "col": 35,
                                "row": 1,
                              },
                              "startPos": {
                                "col": 0,
                                "row": 1,
                              },
                              "type": "node",
                            },
                            {
                              "content": "
",
                              "endPos": {
                                "col": 0,
                                "row": 2,
                              },
                              "startPos": {
                                "col": 35,
                                "row": 1,
                              },
                              "type": "string",
                            },
                          ],
                          "endPos": {
                            "col": 0,
                            "row": 2,
                          },
                          "startPos": {
                            "col": 0,
                            "row": 1,
                          },
                          "type": "node",
                        },
                      ],
                      "endPos": {
                        "col": 0,
                        "row": 2,
                      },
                      "startPos": {
                        "col": 0,
                        "row": 1,
                      },
                      "type": "array",
                    },
                    {
                      "content": "",
                      "endPos": {
                        "col": 0,
                        "row": 2,
                      },
                      "startPos": {
                        "col": 0,
                        "row": 2,
                      },
                      "type": "string",
                    },
                  ],
                  "endPos": {
                    "col": 0,
                    "row": 2,
                  },
                  "startPos": {
                    "col": 0,
                    "row": 0,
                  },
                  "type": "node",
                },
                {
                  "content": "
",
                  "endPos": {
                    "col": 0,
                    "row": 3,
                  },
                  "startPos": {
                    "col": 0,
                    "row": 2,
                  },
                  "type": "string",
                },
              ],
              "endPos": {
                "col": 0,
                "row": 3,
              },
              "startPos": {
                "col": 0,
                "row": 0,
              },
              "type": "node",
            },
            {
              "children": [
                {
                  "children": [
                    {
                      "content": "# ",
                      "endPos": {
                        "col": 2,
                        "row": 3,
                      },
                      "startPos": {
                        "col": 0,
                        "row": 3,
                      },
                      "type": "string",
                    },
                    {
                      "content": "assistant",
                      "endPos": {
                        "col": 11,
                        "row": 3,
                      },
                      "startPos": {
                        "col": 2,
                        "row": 3,
                      },
                      "type": "string",
                    },
                    {
                      "content": ":
",
                      "endPos": {
                        "col": 0,
                        "row": 4,
                      },
                      "startPos": {
                        "col": 11,
                        "row": 3,
                      },
                      "type": "string",
                    },
                    {
                      "children": [
                        {
                          "children": [
                            {
                              "children": [
                                {
                                  "content": "Sure, let me use the list_buffers tool.",
                                  "endPos": {
                                    "col": 39,
                                    "row": 4,
                                  },
                                  "startPos": {
                                    "col": 0,
                                    "row": 4,
                                  },
                                  "type": "string",
                                },
                              ],
                              "endPos": {
                                "col": 39,
                                "row": 4,
                              },
                              "startPos": {
                                "col": 0,
                                "row": 4,
                              },
                              "type": "node",
                            },
                            {
                              "content": "
",
                              "endPos": {
                                "col": 0,
                                "row": 5,
                              },
                              "startPos": {
                                "col": 39,
                                "row": 4,
                              },
                              "type": "string",
                            },
                          ],
                          "endPos": {
                            "col": 0,
                            "row": 5,
                          },
                          "startPos": {
                            "col": 0,
                            "row": 4,
                          },
                          "type": "node",
                        },
                        {
                          "children": [
                            {
                              "children": [
                                {
                                  "children": [
                                    {
                                      "content": "⚙️ Grabbing buffers...",
                                      "endPos": {
                                        "col": 26,
                                        "row": 5,
                                      },
                                      "startPos": {
                                        "col": 0,
                                        "row": 5,
                                      },
                                      "type": "string",
                                    },
                                  ],
                                  "endPos": {
                                    "col": 26,
                                    "row": 5,
                                  },
                                  "startPos": {
                                    "col": 0,
                                    "row": 5,
                                  },
                                  "type": "node",
                                },
                                {
                                  "content": "",
                                  "endPos": {
                                    "col": 26,
                                    "row": 5,
                                  },
                                  "startPos": {
                                    "col": 26,
                                    "row": 5,
                                  },
                                  "type": "string",
                                },
                                {
                                  "content": "",
                                  "endPos": {
                                    "col": 26,
                                    "row": 5,
                                  },
                                  "startPos": {
                                    "col": 26,
                                    "row": 5,
                                  },
                                  "type": "string",
                                },
                              ],
                              "endPos": {
                                "col": 26,
                                "row": 5,
                              },
                              "startPos": {
                                "col": 0,
                                "row": 5,
                              },
                              "type": "node",
                            },
                            {
                              "content": "
",
                              "endPos": {
                                "col": 0,
                                "row": 6,
                              },
                              "startPos": {
                                "col": 26,
                                "row": 5,
                              },
                              "type": "string",
                            },
                          ],
                          "endPos": {
                            "col": 0,
                            "row": 6,
                          },
                          "startPos": {
                            "col": 0,
                            "row": 5,
                          },
                          "type": "node",
                        },
                      ],
                      "endPos": {
                        "col": 0,
                        "row": 6,
                      },
                      "startPos": {
                        "col": 0,
                        "row": 4,
                      },
                      "type": "array",
                    },
                    {
                      "content": "",
                      "endPos": {
                        "col": 0,
                        "row": 6,
                      },
                      "startPos": {
                        "col": 0,
                        "row": 6,
                      },
                      "type": "string",
                    },
                  ],
                  "endPos": {
                    "col": 0,
                    "row": 6,
                  },
                  "startPos": {
                    "col": 0,
                    "row": 3,
                  },
                  "type": "node",
                },
                {
                  "content": "
",
                  "endPos": {
                    "col": 0,
                    "row": 7,
                  },
                  "startPos": {
                    "col": 0,
                    "row": 6,
                  },
                  "type": "string",
                },
              ],
              "endPos": {
                "col": 0,
                "row": 7,
              },
              "startPos": {
                "col": 0,
                "row": 3,
              },
              "type": "node",
            },
          ],
          "endPos": {
            "col": 0,
            "row": 7,
          },
          "startPos": {
            "col": 0,
            "row": 0,
          },
          "type": "array",
        },
        {
          "children": [
            {
              "content": "Stopped (",
              "endPos": {
                "col": 9,
                "row": 7,
              },
              "startPos": {
                "col": 0,
                "row": 7,
              },
              "type": "string",
            },
            {
              "content": "end_turn",
              "endPos": {
                "col": 17,
                "row": 7,
              },
              "startPos": {
                "col": 9,
                "row": 7,
              },
              "type": "string",
            },
            {
              "content": ") [input: ",
              "endPos": {
                "col": 27,
                "row": 7,
              },
              "startPos": {
                "col": 17,
                "row": 7,
              },
              "type": "string",
            },
            {
              "content": "0",
              "endPos": {
                "col": 28,
                "row": 7,
              },
              "startPos": {
                "col": 27,
                "row": 7,
              },
              "type": "string",
            },
            {
              "content": ", output: ",
              "endPos": {
                "col": 38,
                "row": 7,
              },
              "startPos": {
                "col": 28,
                "row": 7,
              },
              "type": "string",
            },
            {
              "content": "0",
              "endPos": {
                "col": 39,
                "row": 7,
              },
              "startPos": {
                "col": 38,
                "row": 7,
              },
              "type": "string",
            },
            {
              "content": "",
              "endPos": {
                "col": 39,
                "row": 7,
              },
              "startPos": {
                "col": 39,
                "row": 7,
              },
              "type": "string",
            },
            {
              "content": "]",
              "endPos": {
                "col": 40,
                "row": 7,
              },
              "startPos": {
                "col": 39,
                "row": 7,
              },
              "type": "string",
            },
          ],
          "endPos": {
            "col": 40,
            "row": 7,
          },
          "startPos": {
            "col": 0,
            "row": 7,
          },
          "type": "node",
        },
        {
          "content": "",
          "endPos": {
            "col": 40,
            "row": 7,
          },
          "startPos": {
            "col": 40,
            "row": 7,
          },
          "type": "string",
        },
      ],
      "endPos": {
        "col": 40,
        "row": 7,
      },
      "startPos": {
        "col": 0,
        "row": 0,
      },
      "type": "node",
    },
  ],
  "endPos": {
    "col": 40,
    "row": 7,
  },
  "startPos": {
    "col": 0,
    "row": 0,
  },
  "type": "node",
}
`;
